### ISSUE TITLE
'bundle exec' loads gemified openssl/io-console from outside the restricted gem set

### ISSUE THREAD

**Issue by rhenium** [bundler/bundler#5235]
`bundle exec` loads io/console and openssl before restricting load path and causes two different versions of the library to be loaded, if the latest installed version is not specified in Gemfile.

At the moment of Ruby 2.4, only io/console and openssl are affected because other stdlibs loaded by `bundle exec` aren't (yet?[1]) converted into default (!= bundled) gems.

Steps to reproduce (with io/console, but can be read as openssl if you use Ruby >= 2.4.0-preview1):

1. Install Ruby 2.3.3 which includes io-console 0.4.5 as a default gem and Bundler 1.13.6.
2. Run `gem install io-console` to install io-console 0.4.6. `gem list` shows `io-console (0.4.6, 0.4.5)`.
3. Setup an app with:

    Gemfile:

    ```
    source "https://rubygems.org"
    ```

    main.rb:

    ```
    puts "before require: #{$LOADED_FEATURES.grep(/console/)}"
    require "io/console"
    puts "after require: #{$LOADED_FEATURES.grep(/console/)}"
    ```

4. Run `bundle exec ruby -w main.rb`. The output shows that two versions are loaded:

    ```
    before require: ["/home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/io-console-0.4.6/lib/io/console.so"]
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old raw
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old raw!
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old cooked
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old cooked!
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old getch
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old echo=
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old echo?
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old noecho
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old winsize
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old winsize=
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old iflush
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old oflush
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old ioflush
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old beep
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old goto
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old cursor
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old cursor=
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old pressed?
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old getpass
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old console
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old getch
    /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so: warning: method redefined; discarding old getpass
    after require: ["/home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/io-console-0.4.6/lib/io/console.so", "/home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/x86_64-linux/io/console.so"]
    ```

[1]: https://bugs.ruby-lang.org/issues/5481

**Comment by segiddins** [comment]
Why is ruby loading the second one, when the first is already in `$LOADED_FEATURES`?

**Comment by rhenium** [comment]
Because the full path of the two console.so are different. I expected only the second one (explicitly `require`ed in main.rb) is loaded, rather than the first one, however.

This example may be better: if I put explicit `gem "io-console", "0.4.5"` to Gemfile, I get the following error:

```
$ bundle exec ruby -e'1'
/home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.13.6/lib/bundler/runtime.rb:40:in `block in setup': You have already activated io-console 0.4.6, but your Gemfile requires io-console 0.4.5. Prepending `bundle exec` to your command may solve this. (Gem::LoadError)
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.13.6/lib/bundler/runtime.rb:25:in `map'
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.13.6/lib/bundler/runtime.rb:25:in `setup'
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.13.6/lib/bundler.rb:102:in `setup'
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/bundler-1.13.6/lib/bundler/setup.rb:10:in `<top (required)>'
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /home/k/.rbenv/versions/2.3.3/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
```

**Comment by segiddins** [comment]
https://github.com/rubygems/rubygems/pull/1674 will be required to fix this, as will a patch to thor to not eagerly load io/console

**Comment by nomoon** [comment]
I've had success working around this by putting the following line near the top of my `Gemfile`:
```ruby
if defined?(RUBY_DESCRIPTION) && RUBY_DESCRIPTION.start_with?("ruby 2.4")
  gem "openssl"
end
```

**Comment by will-in-wi** [comment]
We upgraded bundler on our production systems recently, and 984f64d broke one of our apps. I tracked it down by bisecting bundler from the old version to the new one, running this file each time:

```ruby
puts Time.at(1234).to_datetime.to_s
```

Before this commit, the output was `1969-12-31T19:20:34-05:00`. After this commit, the output is `test.rb:1:in `<main>': undefined method `to_datetime' for 1969-12-31 19:20:34 -0500:Time (NoMethodError)`.

Technically, this exposed a bug in our application, which we've now fixed. I just wanted to note that this broke compatibility in a subtle way. I don't think any changes in Bundler are needed.

Thanks!
